lib/services/local/generic/generic153_test.go

```diff
@@ -0,0 +1,277 @@
+type testResource153 struct {
+	Metadata *headerv1.Metadata
+}
+
+func (t *testResource153) GetMetadata() *headerv1.Metadata {
+	return t.Metadata
+}
+
+var _ types.ResourceMetadata = (*testResource153)(nil)
+var _ Resource = (*resourceMetadataAdapter[*testResource153])(nil)
+
+func newTestResource153(name string) *testResource153 {
+	tr := &testResource153{
+		Metadata: &headerv1.Metadata{
+			Name: name,
+		},
+	}
+	return tr
+}
+
+// marshalResource153 marshals a generic resource.
+func marshalResource153(resource *testResource153, opts ...services.MarshalOption) ([]byte, error) {
+	return utils.FastMarshal(resource)
+}
+
+// unmarshalResource153 unmarshals a generic resource.
+func unmarshalResource153(data []byte, opts ...services.MarshalOption) (*testResource153, error) {
+	if len(data) == 0 {
+		return nil, trace.BadParameter("missing resource data")
+	}
+	cfg, err := services.CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var h types.ResourceHeader
+	if err := utils.FastUnmarshal(data, &h); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var r testResource153
+	if err := utils.FastUnmarshal(data, &r); err != nil {
+		return nil, trace.BadParameter(err.Error())
+	}
+
+	if r.Metadata == nil {
+		r.Metadata = &headerv1.Metadata{}
+	}
+
+	if cfg.ID != 0 {
+		//nolint:staticcheck // SA1019. Deprecated, but still needed.
+		r.Metadata.Id = cfg.ID
+	}
+	if cfg.Revision != "" {
+		r.Metadata.Revision = cfg.Revision
+	}
+	if !cfg.Expires.IsZero() {
+		r.Metadata.Expires = timestamppb.New(cfg.Expires)
+	}
+	return &r, nil
+}
+
+// TestGenericCRUD153 tests backend operations with the generic service.
+func TestGenericCRUD153(t *testing.T) {
+	ctx := context.Background()
+
+	memBackend, err := memory.New(memory.Config{
+		Context: ctx,
+		Clock:   clockwork.NewFakeClock(),
+	})
+	require.NoError(t, err)
+
+	const backendPrefix = "generic_prefix"
+
+	service, err := NewService153[*testResource153](memBackend,
+		"generic resource",
+		backendPrefix,
+		marshalResource153,
+		unmarshalResource153)
+	require.NoError(t, err)
+
+	// Create a couple test resources.
+	r1 := newTestResource153("r1")
+	r2 := newTestResource153("r2")
+
+	// sanity check: marshal/unmarshal round trips
+	r1data, err := marshalResource153(r1)
+	require.NoError(t, err)
+	r1clone, err := unmarshalResource153(r1data)
+	require.NoError(t, err)
+	require.Equal(t, r1, r1clone)
+
+	// Initially we expect no resources.
+	out, nextToken, err := service.ListResources(ctx, 200, "")
+	require.NoError(t, err)
+	require.Empty(t, nextToken)
+	require.Empty(t, out)
+
+	// Create both resources.
+	r1, err = service.CreateResource(ctx, r1)
+	require.NoError(t, err)
+	r2, err = service.CreateResource(ctx, r2)
+	require.NoError(t, err)
+
+	require.NotEmpty(t, r1.GetMetadata().GetRevision())
+	require.NotEmpty(t, r2.GetMetadata().GetRevision())
+	require.NotEqual(t, r1.GetMetadata().GetRevision(), r2.GetMetadata().GetRevision())
```

Comment:
Is this always guaranteed to be true? Why can't different resources have the same revision?

Reply:
This is from copied from a very recent change: 
- https://github.com/gravitational/teleport/pull/36183

@rosstimothy do you remember your rationale?